### ISSUE TITLE
Test transaction/sandboxing lost if worker exits

### ISSUE THREAD

**Issue by fishcakez** [elixir-lang/ecto#609]
If the (single) worker exits when using a test transaction the replacement worker does not sandbox its connection.

Using the simple example after changing the size to 1:
```elixir
:ok = Ecto.Adapters.SQL.begin_test_transaction(Simple.Repo)
poolboy = elem(Simple.Repo.__pool__, 0)
worker = :poolboy.checkout(poolboy, true, :infinity) 
true = :sys.get_state(worker).sandbox
Process.exit(worker, :kill)
worker = :poolboy.checkout(poolboy, true, :infinity)
false = :sys.get_state(worker).sandbox
```

**Comment by ericmj** [comment]
@josevalim Can we just start a new transaction if `sandbox` is `false`?

**Comment by josevalim** [comment]
If sandbox is true you mean?

**Comment by ericmj** [comment]
The restarted worker should have `sandbox: false` since it's a fresh worker. If `restart_test_transaction` is called when `sandbox: false` we just ignore the call today, why not start a new transaction instead?

**Comment by fishcakez** [comment]
Can't queries be made in between the exit (i.e. to the replacement worker) and a `restart_test_transaction` call?

**Comment by josevalim** [comment]
@fishcakez it can but it will be wrong anyway (because they will be in the wrong transaction).

**Comment by fishcakez** [comment]
Right, it is wrong transaction or rather there is no active transaction at that point. This means sandboxing isn't happening and when `restart_test_transaction` is called the sandbox is created after the queries between the exit and the restart. This means there is opportunity for sandboxing not to occur and to create the wrong sandbox.

**Comment by ericmj** [comment]
> Can't queries be made in between the exit (i.e. to the replacement worker) and a restart_test_transaction call?

I think this can only happen if you run async tests - which you shouldn't.

**Comment by fishcakez** [comment]
A single test might include multiple processes.

**Comment by ericmj** [comment]
Good point, the worker needs to know it is running during testing when it restarts.

**Comment by josevalim** [comment]
Fixed in #622. restart will start a new one if one does not yet exist.
